### ISSUE TITLE
Batch fetch not working as expected

### ISSUE THREAD

**Issue by diorcety** [nats-io/nats.py#617]
### Observed behavior

Stream S is created
Client A publishes X messages on stream S
Client B pulls X+Y messages on stream (timeout 10 seconds)
Client A publishes Y messages on stream S after 5 seconds
After 10 seconds, only X messages are received by client B

Here the python script in order to reproduce the isse
```
import asyncio

import nats


data = [73, 27]
subject = "AA.BB"
stream1 = "test1"
stream2 = "test1"
consumer = None #"CONSUMER"

fetch_one = False


async def publish(js, msg, delay=0):
    if delay > 0:
        await asyncio.sleep(delay)
    print(f"Publish msg: {msg}")
    await js.publish(subject, msg.encode(), stream=stream1)


async def main():
    nc = await nats.connect("localhost")
    nc2 = await nats.connect("localhost")

    # Create JetStream context.
    js = nc.jetstream()
    js2 = nc.jetstream()

    if stream1 == stream2:
        await js.add_stream(name=stream1, subjects=[subject])

    for i in range(data[0]):
        await publish(js2, f'Hello {i}')

    event = asyncio.gather(*[publish(js2, f'World {i}', 5) for i in range(data[1])])

    sub = await js.pull_subscribe(subject, consumer, stream2)

    # Fetch and ack messages from consumer.
    for i in range(data[0] + data[1] if fetch_one else 1):
        _, msgs = await asyncio.gather(event, sub.fetch(1 if fetch_one else data[0] + data[1], timeout=10))
        #random.shuffle(msgs)
        print(len(msgs))
        for msg in msgs:
            await msg.ack()
            print(msg)

    await nc.close()
    await nc2.close()

if __name__ == '__main__':
    asyncio.run(main())
```

[dump.txt](https://github.com/user-attachments/files/17244198/dump.txt)

### Expected behavior

Stream S is created
Client A publishes X messages on stream S
Client B pulls X+Y messages on stream (timeout 10 seconds)
Client A publishes Y messages on stream S after 5 seconds
Client B pulls X+ Y messages

### Server and client version

nats-server: 2.10.21
nats-py-2.9.0

**Comment by diorcety** [comment]
In the documentation 
https://docs.nats.io/reference/reference-protocols/nats_api_reference

> Here we see a second format of the Pull request that will not store the request on the queue at all but when there are no messages to deliver will send a nil bytes message with a Status header of 404, this way you can know when you reached the end of the stream for example. A 409 is returned if the Consumer has reached MaxAckPending limits.


Here an example
```
> nats pub metrics.demo "hello world"
16:25:35 Published 11 bytes to "metrics.demo"
> nats req --replies=20 '$JS.API.CONSUMER.MSG.NEXT.metrics_cloud.A' '{"no_wait": true, "batch": 10}'
16:25:36 Sending request on "$JS.API.CONSUMER.MSG.NEXT.metrics_cloud.A"
16:25:36 Received with rtt 554.131µs
16:25:36 Nats-Stream-Source: metrics_latest 1 > >

hello world


16:25:36 Received with rtt 701.639µs
16:25:36 Nats-Pending-Messages: 9
16:25:36 Nats-Pending-Bytes: 0
16:25:36 Status: 408
16:25:36 Description: Request Timeout

nil body

```

It really seems there is an issue with the server. The request timeout doesn't seem to be the correct reply

**Comment by diorcety** [comment]
@wallyqs You may have to transfer the message back

**Comment by diorcety** [comment]
With this patch https://github.com/diorcety/nats-server/commit/7c5906bca2c9297fcaca5263b932b61f95c46200

```
> nats pub metrics.demo "hello world"
16:30:02 Published 11 bytes to "metrics.demo"
> nats req --replies=20 '$JS.API.CONSUMER.MSG.NEXT.metrics_cloud.A' '{"no_wait": true, "batch": 10}'
16:30:03 Sending request on "$JS.API.CONSUMER.MSG.NEXT.metrics_cloud.A"
16:30:03 Received with rtt 560.134µs
16:30:03 Nats-Stream-Source: metrics_latest 1 > >

hello world


16:30:03 Received with rtt 693.042µs
16:30:03 Nats-Pending-Messages: 9
16:30:03 Nats-Pending-Bytes: 0
16:30:03 Status: 404
16:30:03 Description: No Messages

nil body
```

**Comment by diorcety** [comment]
Or this is the expected behavior (no_wait, with no more message meaning timeout), then the issue is in the python implementation

**Comment by diorcety** [comment]
@wallyqs 
the example using nats cli , show the issue. nats.py make a first no_wait request. If you didn't fullfill the number of message which the requested one, the last message from the server is currently a 408 timeout.
The nats.py expects a 404 No Messages and not a 408. 
Here a fix in python library
https://github.com/diorcety/nats.py/commit/0530a9ceb43e904181e715b23ed652dcf5327f1a

**Comment by wallyqs** [comment]
thanks @diorcety, will take a look it sounds like it might be this issue: https://github.com/nats-io/nats-server/issues/5373

**Comment by wallyqs** [comment]
@diorcety we will fix the issue in the server for v2.11, but do you want to send PR for the nats.py fix to include as well?
